File: example/cs/tmr/ResolutionChangeExample.cs
//*****************************************************************************************************************
//!
//! @file    ResolutionChangeExample.cs
//! @par        Character Code: UTF-8N
//! @par        Return Code: LF
//!
//! @brief   \~japanese Close／再Openによる2つのStream ProfileでのFrame取得.
//! @brief   \~english  Capturing one frame at each of two stream profiles by closing and reopening.
//!
//! @details
//!     \~japanese
//!     @n 引数順序は arguments[0] がNative Library Pathである。追加引数は取らない。
//!     @n 解像度はOpen時に固定される。変更にはCloseと別Profileでの再Openが必要である。
//!     @n Cameraが無い環境でも失敗しない。列挙結果が空であることを報告して終了する。
//!     @n C++版はBMPを保存するが、BindingにImage Writerが無い。代わりに両Frameの寸法と
//!        Byte数を出力する。
//!     \~english
//!     @n Argument order: arguments[0] is the native library path. The sample takes no further
//!        arguments.
//!     @n The resolution is fixed at open, so changing it means closing the session and reopening it
//!        with a second stream profile.
//!     @n A machine with no camera does not fail: the sample reports the empty enumeration and
//!        stops.
//!     @n The C++ counterpart saves each capture as a BMP. The binding exposes no image writer, so
//!        this sample reports each frame's size and byte count instead.
//!
//! @date
//!   Sep-06, 2026   Create New.
//!
//! @copyright
//!   Copyright (C) 2026 WapitiStew. All rights reserved.
//*****************************************************************************************************************

using System;
using System.Collections.Generic;
using System.Linq;
using WapitiStew.Wse;

/// <summary>One frame at each of two stream profiles of the same camera.</summary>
public static class ResolutionChangeExample
{
    /// <summary>How long one frame read may take.</summary>
    // Both captures here are the first frame of a fresh session, which is the slowest kind: the
    // device has just been opened and is still negotiating the stream. Two seconds absorbs that
    // where a deadline near the frame interval would report TimedOut on a healthy camera.
    private static readonly TimeSpan ReadTimeout = TimeSpan.FromSeconds(2);

    /// <summary>Captures at the first profile, reopens at a different one, and captures again.</summary>
    /// <param name="arguments">arguments[0] is the optional path to the native library file.</param>
    /// <returns>Zero on success and on every machine without a usable camera.</returns>
    public static int Main(string[] arguments)
    {
        if (arguments.Length > 0)
        {
            WseRuntime.SetNativeLibraryPath(arguments[0]);
        }

        using var runtime = new WseRuntime();
        if (!runtime.Info().Components.HasTmr)
        {
            Console.Error.WriteLine("This WSE build does not include the Tmr component.");
            return 1;
        }

        // Every enumerated device is a native description the caller owns, which is why the whole
        // body sits in a try whose finally disposes each one.
        IReadOnlyList<CameraDevice> devices = WebCamera.Enumerate();
        try
        {
            // An empty list is the ordinary answer on a machine with no camera, not a failure.
            if (devices.Count == 0)
            {
                Console.WriteLine("No USB/UVC web camera is attached.");
                return 0;
            }

            // One device is enough here, because the point is two profiles of the same camera
            // rather than a choice between cameras. It stays owned by the list.
            CameraDevice device = devices[0];
            Console.WriteLine($"device: {device.DisplayName}");

            // Two profiles with different frame sizes; both have to offer a usable output format.
            // Only the frame size is compared, because a profile that differs solely in frame rate
            // or in native format would not demonstrate a resolution change. The whole search runs
            // inside one capability scope so the native capability set is released before the
            // device is opened.
            CameraStreamConfiguration first = default;
            CameraStreamConfiguration second = default;
            bool firstFound = false;
            bool secondFound = false;
            using (CameraCapability capability = WebCamera.Capabilities(device))
            {
                foreach (CameraStreamProfile profile in capability.StreamProfiles())
                {
                    if (profile.OutputFormats.Count == 0) continue;
                    if (!firstFound)
                    {
                        first = new CameraStreamConfiguration(
                            profile.NativeFormat, profile.OutputFormats[0], true);
                        firstFound = true;
                        continue;
                    }
                    bool differentSize =
                        profile.NativeFormat.Width != first.NativeFormat.Width ||
                        profile.NativeFormat.Height != first.NativeFormat.Height;
                    if (!differentSize) continue;

                    // Keep the first output format when the second profile can also produce it.
                    // Holding the format steady leaves the frame size as the only difference
                    // between the two captures, which is what the printed comparison is about.
                    CameraPixelFormat output = profile.OutputFormats.Contains(first.OutputFormat)
                        ? first.OutputFormat
                        : profile.OutputFormats[0];
                    second = new CameraStreamConfiguration(profile.NativeFormat, output, true);
                    secondFound = true;
                    break;
                }
            }

            if (!firstFound)
            {
                Console.WriteLine("No enumerated camera has a usable stream profile.");
                return 0;
            }

            // One camera object serves both sessions. Close ends a session without ending the
            // object, so the same instance opens again with the second configuration; only Dispose
            // releases the native camera, and the using does that once at the end.
            using var camera = new WebCamera();
            try
            {
                Console.WriteLine($"first resolution: {Describe(first.NativeFormat)}");
                string firstResult = CaptureOnce(camera, device, first);
                Console.WriteLine($"first capture: {firstResult}");

                if (!secondFound)
                {
                    Console.WriteLine(
                        "the camera advertises only one resolution; nothing to change");
                    return 0;
                }

                Console.WriteLine($"second resolution: {Describe(second.NativeFormat)}");
                string secondResult = CaptureOnce(camera, device, second);
                Console.WriteLine($"second capture: {secondResult}");
                Console.WriteLine(
                    $"captured both resolutions: {Describe(first.NativeFormat)}"
                    + $" and {Describe(second.NativeFormat)}");
                return 0;
            }
            catch (WseException failure)
            {
                Console.Error.WriteLine(
                    $"ERROR: the camera session failed. category={failure.Category}"
                    + $" code={failure.Code} native={failure.NativeCode} message={failure.Message}");
                return 3;
            }
        }
        finally
        {
            // Enumerate handed over one native description per device, so each is released here
            // whichever way the body above left.
            foreach (CameraDevice item in devices)
            {
                item.Dispose();
            }
        }
    }

    /// <summary>Opens the device with one configuration, reads a frame, and closes again.</summary>
    /// <param name="camera">The camera owner reused across both sessions.</param>
    /// <param name="device">The device to open.</param>
    /// <param name="configuration">Native format, output format, and conversion policy.</param>
    /// <returns>A one-line description of the captured frame.</returns>
    private static string CaptureOnce(
        WebCamera camera, CameraDevice device, CameraStreamConfiguration configuration)
    {
        camera.Open(device, configuration);
        camera.Start();
        try
        {
            // The frame owns native pixels; the using releases them once the description has been
            // formatted. ToArray copies the bytes out, so its length is the real payload size and
            // reflects the resolution just opened.
            using CameraFrame frame = camera.ReadFrame(ReadTimeout);
            return $"{frame.Description.Width}x{frame.Description.Height}"
                + $" format={frame.Description.PixelFormat} stride={frame.Description.RowStride}"
                + $" bytes={frame.ToArray().Length}";
        }
        finally
        {
            // Close ends the session and is idempotent, and it releases the device before the next
            // open. It does not end the camera object: this same instance is opened again by the
            // second call, which is the whole mechanism the sample demonstrates.
            camera.Stop();
            camera.Close();
        }
    }

    /// <summary>Formats a stream format as its pixel dimensions.</summary>
    /// <param name="format">The format to describe.</param>
    /// <returns>The formatted size.</returns>
    private static string Describe(CameraFormat format) => $"{format.Width}x{format.Height}";
}

File: example/java/tmr/ResolutionChangeExample.java
import io.wapitistew.wse.CameraCapability;
import io.wapitistew.wse.CameraDevice;
import io.wapitistew.wse.CameraFormat;
import io.wapitistew.wse.CameraFrame;
import io.wapitistew.wse.CameraPixelFormat;
import io.wapitistew.wse.CameraStreamConfiguration;
import io.wapitistew.wse.CameraStreamProfile;
import io.wapitistew.wse.WebCamera;
import io.wapitistew.wse.WseException;
import io.wapitistew.wse.WseRuntime;

/**
 * Captures one frame at the first advertised stream profile, closes the session, reopens the camera
 * at a profile with a different frame size, captures again, and reports both sizes.
 *
 * <p>The resolution is fixed when the session opens, so changing it means closing and reopening.
 *
 * <p>Hardware-free by design: no camera, no usable stream profile, or a camera that advertises only
 * one resolution is reported in one plain sentence and the sample returns normally. Build with
 * {@code WSE_BUILD_TMR=ON} and {@code WSE_BUILD_JAVA_BINDING=ON}.
 *
 * <p>Arguments: none. Like every other WSE Java sample, the native library is located through the
 * {@code -Dwse.runtime.path=...} system property rather than a command-line argument, so this
 * sample defines no arguments of its own. That load is by absolute path, which on Windows does not
 * add the containing directory to the dependency search path, so the engine directory must also be
 * on {@code PATH}.
 *
 * <p>One difference from the C++ counterpart {@code example/cpp/tmr/resolution_change.cpp}: it
 * saves each capture as a BMP, and the Java binding exposes no image encoder, so the nearest bound
 * thing is done here and each capture is reported by its frame descriptor. Like the C++ sample,
 * this one reuses a single {@code WebCamera} across both sessions; {@code closeCamera()} ends a
 * session and leaves the camera reopenable, while {@code close()} releases the native object.
 */
public final class ResolutionChangeExample {
    private ResolutionChangeExample() {}

    /** Deadline for one frame. 2000 ms absorbs the first-frame delay after start(); each capture
     * here is the first frame of a freshly opened session, which is the slowest one a session ever
     * delivers, so a deadline near the frame interval would time out on both captures. */
    private static final int READ_TIMEOUT_MILLISECONDS = 2000;

    private static void reportFailure(String operation, WseException failure) {
        // Branch on category and code, never on the message text.
        System.out.println(operation + " failed: category=" + failure.category()
                + " code=" + failure.code() + " native=" + failure.nativeCode()
                + " message=" + failure.getMessage());
    }

    private static String describeFormat(CameraFormat format) {
        return format.width() + "x" + format.height();
    }

    /**
     * Opens one session on the given camera, captures one frame, and returns its description.
     *
     * <p>Any failure leaves as a {@code WseException} for the caller to report, and the session is
     * ended either way. The frame's own dimensions are read rather than the requested ones, because
     * the descriptor is what proves the device honoured the profile.
     */
    private static String captureOnce(
            WebCamera camera, CameraDevice device, CameraStreamConfiguration configuration) {
        camera.open(device, configuration);
        try {
            camera.start();
            // The frame owns its pixel buffer; the block frees it before this method returns, so
            // nothing native outlives the session it came from.
            try (CameraFrame frame = camera.readFrame(READ_TIMEOUT_MILLISECONDS)) {
                return frame.width() + "x" + frame.height()
                        + " format=" + frame.pixelFormat()
                        + " sequence=" + frame.sequence()
                        + " bytes=" + frame.data().buffer().remaining();
            }
        } finally {
            // Ends this session only. The camera object stays usable, so the next profile opens
            // on the same instance. It also stops streaming, and calling it twice is safe, so it
            // is correct on the failure path as well - which is why it sits in a finally rather
            // than after the capture.
            camera.closeCamera();
        }
    }

    public static void main(String[] arguments) {
        if (!WseRuntime.info().hasTmr()) {
            System.out.println("This WSE build does not include the Tmr component.");
            return;
        }

        // Enumeration is static and needs no session; an empty list means no camera, which is a
        // reportable state rather than a failure so the sample runs on a machine without one.
        CameraDevice[] devices = WebCamera.enumerate();
        if (devices.length == 0) {
            System.out.println("No USB/UVC web camera is attached.");
            return;
        }

        // The whole capability is kept here, not just one profile: the second resolution is chosen
        // from the same advertised list further down.
        CameraDevice device = null;
        CameraCapability capability = null;
        for (CameraDevice candidate : devices) {
            CameraCapability candidateCapability = WebCamera.capabilities(candidate);
            if (candidateCapability.streamProfiles().length != 0
                    && candidateCapability.streamProfiles()[0].outputFormats().length != 0) {
                device = candidate;
                capability = candidateCapability;
                break;
            }
        }
        if (device == null || capability == null) {
            System.out.println("No enumerated camera has a usable stream profile.");
            return;
        }
        System.out.println("camera: " + device.displayName());

        // The trailing true allows conversion from the native format into the requested output;
        // with false the device itself would have to produce that output, which narrows how many
        // profiles are usable and would work against picking a second one below.
        CameraStreamProfile firstProfile = capability.streamProfiles()[0];
        CameraStreamConfiguration first = new CameraStreamConfiguration(
                firstProfile.nativeFormat(), firstProfile.outputFormats()[0], true);

        // A second profile with a different frame size; keep the same output format when it is
        // offered so only the resolution differs between the two captures.
        CameraStreamConfiguration second = null;
        for (CameraStreamProfile profile : capability.streamProfiles()) {
            boolean differentSize =
                    profile.nativeFormat().width() != first.nativeFormat().width()
                    || profile.nativeFormat().height() != first.nativeFormat().height();
            if (!differentSize || profile.outputFormats().length == 0) continue;
            // A profile that offers the first capture's output format is preferred so the two
            // descriptors differ in size alone; otherwise the format changes too and the
            // comparison at the end is less direct.
            CameraPixelFormat output = profile.supportsOutput(first.outputFormat())
                    ? first.outputFormat()
                    : profile.outputFormats()[0];
            second = new CameraStreamConfiguration(profile.nativeFormat(), output, true);
            break;
        }

        // One camera object serves both sessions: closeCamera() ends a session and leaves the
        // object reopenable, and try-with-resources releases it once for good at the end.
        // Getting this pair the wrong way round is the mistake this sample exists to prevent:
        // close() is terminal, so calling it between the two captures would leave the second open
        // to fail on a released camera.
        try (WebCamera camera = new WebCamera()) {
            try {
                System.out.println("first profile: " + describeFormat(first.nativeFormat())
                        + " at " + first.nativeFormat().framesPerSecond() + " fps"
                        + " output=" + first.outputFormat());
                System.out.println("first capture: " + captureOnce(camera, device, first));
            } catch (WseException failure) {
                reportFailure("capture at " + describeFormat(first.nativeFormat()), failure);
                return;
            }

            // A fixed-resolution camera is a real device, not a broken one, so this returns
            // normally after the first capture has already been reported.
            if (second == null) {
                System.out.println("The camera advertises only one resolution; nothing to change.");
                return;
            }

            try {
                System.out.println("second profile: " + describeFormat(second.nativeFormat())
                        + " at " + second.nativeFormat().framesPerSecond() + " fps"
                        + " output=" + second.outputFormat());
                System.out.println("second capture: " + captureOnce(camera, device, second));
                System.out.println("captured both resolutions: "
                        + describeFormat(first.nativeFormat()) + " and "
                        + describeFormat(second.nativeFormat()));
            } catch (WseException failure) {
                reportFailure("capture at " + describeFormat(second.nativeFormat()), failure);
            }
        }
    }
}
